src/wit_gen.rs

```diff
@@ -20,63 +22,68 @@ impl<F> StepInstance<F> {
     }
 }
 
-impl<F: Eq + Hash> WitnessGenContext<F> for StepInstance<F> {
-    fn assign(&mut self, lhs: Queriable<F>, rhs: F) {
+impl<F: Eq + Hash> StepInstance<F> {
+    /// Takes a `Queriable` object representing the witness column (lhs) and the value (rhs) to be
+    /// assigned.
+    pub fn assign(&mut self, lhs: Queriable<F>, rhs: F) {
         self.assignments.insert(lhs, rhs);
     }
 }
 
 pub type Witness<F> = Vec<StepInstance<F>>;
 
+#[derive(Debug, Default)]
 pub struct TraceWitness<F> {
     pub step_instances: Witness<F>,
     pub height: usize,
 }
 
-impl<F> Default for TraceWitness<F> {
-    fn default() -> Self {
-        Self {
-            step_instances: Default::default(),
-            height: Default::default(),
-        }
-    }
-}
-
-pub struct GenericTraceContext<'a, F, StepArgs> {
-    step_types: &'a HashMap<u32, Rc<StepType<F, StepArgs>>>,
-
+#[derive(Debug, Default)]
+pub struct TraceContext<F> {
     witness: TraceWitness<F>,
 }
 
-impl<'a, F, StepArgs> GenericTraceContext<'a, F, StepArgs> {
-    pub fn new(step_types: &'a HashMap<u32, Rc<StepType<F, StepArgs>>>) -> Self {
-        Self {
-            step_types,
-            witness: TraceWitness::default(),
-        }
-    }
-
+impl<F> TraceContext<F> {
     pub fn get_witness(self) -> TraceWitness<F> {
         self.witness
     }
 }
 
-impl<'a, F: Eq + Hash, StepArgs> TraceContext<StepArgs> for GenericTraceContext<'a, F, StepArgs> {
-    fn add(&mut self, step: &StepTypeHandler, args: StepArgs) {
-        let step = Rc::clone(
-            self.step_types
-                .get(&step.uuid())
-                .expect("step type not found"),
-        );
-
+impl<F> TraceContext<F> {
+    pub fn add<Args, WG: Fn(&mut StepInstance<F>, Args) + 'static>(
+        &mut self,
+        step: &StepTypeWGHandler<F, Args, WG>,
+        args: Args,
+    ) {
```

Comment:
What's the purpose of having `StepTypeWGHandler` and storing `wg` lambda there instead of storing `wg` in `StepType`?

Reply:
That is the whole point. StepType and Circuit and ir::Circuit should not be parametrized by StepArgs.